### PR TITLE
Graph type builder + hierarchy cleanup

### PR DESCRIPTION
**Pull request by d-michail**
Added a graph builder (using the builder pattern) in order to be able to create graphs without remembering the graph class hierarchy. Called it `GraphTypeBuilder` since we already have a `GraphBuilder`.

This is a long overdue change which allows someone to create a graph like: 
```
Graph<Integer, DefaultEdge> graph = 
   GraphTypeBuilder.<Integer, DefaultEdge> directed()
.allowingMultipleEdges(true).allowingSelfLoops(true).edgeClass(DefaultEdge.class).build();
```

You can also create a graph which has the same type as another graph, like: 

```
Graph<Integer, DefaultEdge> graph = new Pseudograph<>(DefaultEdge.class);
Graph<Integer, DefaultEdge> newGraph = GraphTypeBuilder.asGraph(graph).build();
```

I also did some minor changes in the class hierarchy which we forgot to do when we changed the graph interfaces. Now that we do not have any interfaces called "DirectedGraph" or "UndirectedGraph", we can remove the "Default" prefix from the graph classes. So `DefaultDirectedGraph` is now `DirectedGraph`. I also added an `UndirectedGraph` class to complete the hierarchy.

I also cleaned up a few minor issues that Eclipse was complaining about.

### CODE REVIEW

**Comment by jbduncan** [comment]
As someone who's worked with Guava's `com.google.common.graph.*` types, where there exists a very similar graph builder class, I very much love the idea of this PR. :)

**Comment by jkinable** [comment]
Could you add the examples you state in the intro of this PR to the class description of GraphTypeBuilder? Having such examples directly in the javadoc makes it super easy to use the class (see for instance the javadoc in the standard String class).

**Comment by d-michail** [comment]
Nice suggestion, I added some examples of use. 

Other changes: 
 - I changed the method names as suggested by @jsichi 
 - I also added the term "parallel" in the javadoc comments where we mention multiple edges.

**Comment by d-michail** [comment]
@jsichi I took back the removal of the "Default" prefix for the graph implementations. I noted in the Javadoc that this is the default implementation of a directed graph that the library provides. I also added the undirected variants of such classes, for example "DefaultUndirectedGraph".

I also took some time improving a little bit the Javadocs inside the graph package.

File: jgrapht-core/src/main/java/org/jgrapht/graph/builder/GraphTypeBuilder.java [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+
+    /**
+     * Create a graph type builder which will create a graph with the same type as the one provided.
+     * 
+     * @param type the graph type
+     * @return the graph type builder
+     * @param <V> the graph vertex type
+     * @param <E> the graph edge type
+     */
+    public static <V, E> GraphTypeBuilder<V, E> asGraphType(GraphType type)
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/graph/builder/GraphTypeBuilder.java]: The method naming convention "asX" usually pairs with a conversion to type X.  So I suggest "forGraphType" and "forGraph" instead.

**Comment by jsichi** [comment]
Thanks!

**Pull request merged**